File: MADVCamera/CameraManager/DATAConnectManager.m
//
//  DATAConnectManager.m
//  Madv360_v1
//
//  Created by 张巧隔 on 16/9/7.
//  Copyright © 2016年 Cyllenge. All rights reserved.
//

#import "DATAConnectManager.h"
#import "SocketHelper.h"
#import <sys/socket.h>
#import "NSRecursiveCondition.h"

// Handler Messages :
#define MSG_STATE_CHANGED  10
#define MSG_RECEIVE_ERROR  12
#define MSG_RECEIVE_TIMEOUT  13
#define MSG_RECEIVER_EMPTIED  14

//连接重试次数
#define SOCKET_RETRY_TIMES_LIMIT  2
#define RECEIVE_TIMEOUT_LIMIT  10000
#define BUFFER_SIZE  1048576

@implementation DataReceiver

@synthesize recentReceiveTime;

- (instancetype) init {
    if (self = [super init])
    {
        self.recentReceiveTime = -1;
    }
    return self;
}

- (void) finish {
    [[DATAConnectManager sharedInstance] removeDataReceiver:self];
}

- (int) didReceiveData:(UInt8*)data offset:(int)offset length:(int)length {
    self.recentReceiveTime = [[NSDate date] timeIntervalSince1970] * 1000;
    return [self onReceiveData:data offset:offset length:length];
}

- (int) onReceiveData:(UInt8*)data offset:(int)offset length:(int)length {return 0;}

- (void) onError:(int)error errMsg:(NSString*)errMsg {}

- (BOOL) isFinished {return YES;}

@end

@interface DATAConnectManager ()
{
    NSRecursiveCondition* _cond;
    
    DataReceiver* _currentDataReceiver;
    
    NSString* _serverIP;
    int _serverPort;
    
    int _connTimeout;
    
    DataSocketState _connState;
    
    int _socket;
    
    Byte* _receiveBuffer;
    
    NSThread* _timeoutCheckThread;
    NSThread* _receiveDataThread;
    
    NSMutableArray<id<DataConnectionObserver> >* _observers;
}

@end

@implementation DATAConnectManager

+ (instancetype) sharedInstance {
    static dispatch_once_t once;
    static DATAConnectManager* singleton = nil;
    dispatch_once(&once, ^{
        singleton = [[DATAConnectManager alloc] init];
    });
    return singleton;
}

- (void) dealloc {
    if (_receiveBuffer) free(_receiveBuffer);
}

- (instancetype) init {
    if (self = [super init])
    {
        _cond = [[NSRecursiveCondition alloc] init];
        
        _serverIP = AMBA_CAMERA_IP;
        _serverPort = AMBA_CAMERA_DATA_PORT;
        _connTimeout = AMBA_CAMERA_TIMEOUT;
        
        _receiveBuffer = (Byte*)malloc(BUFFER_SIZE);
        
        _observers = [[NSMutableArray alloc] init];
        _connState = DataSocketStateNotReady;
    }
    return self;
}

- (DataSocketState) state {
    return _connState;
}

- (void) setState:(DataSocketState)state {
    [_cond lock];
    {
        int oldState = _connState;
        [self willChangeValueForKey:@"state"];
        _connState = state;
        [self didChangeValueForKey:@"state"];
        [_cond broadcast];
        
        [self notifyStateChanged:state oldState:(DataSocketState)oldState object:self];
    }
    [_cond unlock];
}

- (int) waitForState:(int)stateCombo {
    [_cond lock];
    {
        while ((stateCombo != 0 || _connState != 0) && ((stateCombo & _connState) == 0))
        {
            [_cond wait];
        }
    }
    [_cond unlock];
    return _connState;
}

- (void) addObserver:(id<DataConnectionObserver>)observer {
    [_observers addObject:observer];
}

- (void) removeObserver:(id<DataConnectionObserver>)observer {
    [_observers removeObject:observer];
}

- (DataReceiver*) removeDataReceiver {
    return [self removeDataReceiver:_currentDataReceiver];
}

- (DataReceiver*) removeDataReceiver:(DataReceiver *)receiver {
    DataReceiver* ret = nil;
    [_cond lock];
    {
        if (_currentDataReceiver == receiver)
        {
            ret = _currentDataReceiver;
            _currentDataReceiver = nil;
            [_cond broadcast];
            
            [self sendMessage:MSG_RECEIVER_EMPTIED arg1:0 arg2:0 object:nil];
        }
    }
    [_cond unlock];
    return ret;
}

- (void) setDataReceiver:(DataReceiver *)dataReceiver {
    [_cond lock];
    {
        dataReceiver.recentReceiveTime = [[NSDate date] timeIntervalSince1970] * 1000;
        _currentDataReceiver = dataReceiver;
        [_cond broadcast];
    }
    [_cond unlock];
}

- (DataReceiver*) dataReceiver {
    return _currentDataReceiver;
}

- (void) openConnection {
    switch (self.state)
    {
        case DataSocketStateConnecting:
        case DataSocketStateReady:
            return;
        case DataSocketStateDisconnecting:
        {
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                [self waitForState:DataSocketStateNotReady];
                self.state = DataSocketStateConnecting;
                [self open];
            });
        }
            break;
        case DataSocketStateNotReady:
        {
            self.state = DataSocketStateConnecting;
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                [self open];
            });
        }
            break;
        default:
            break;
    }
}

- (void) closeConnection {
    switch (self.state)
    {
        case DataSocketStateDisconnecting:
        case DataSocketStateNotReady:
            return;
        case DataSocketStateConnecting:
        {
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                [self waitForState:DataSocketStateReady];
                self.state = DataSocketStateDisconnecting;
                [self close:DataSocketStateNotReady];
            });
        }
            break;
        case DataSocketStateReady:
        {
            self.state = DataSocketStateDisconnecting;
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                [self close:DataSocketStateNotReady];
            });
        }
            break;
        default:
            break;
    }
}

- (BOOL) isConnected {
    return ((DataSocketStateReady == self.state)
            && _receiveDataThread
            && _receiveDataThread.isExecuting);
}

- (void) open {
    __block int retryTimes = 0;
    __block BOOL success = NO;
    while (retryTimes < SOCKET_RETRY_TIMES_LIMIT && self.state != DataSocketStateDisconnecting)
    {
        connectServer(_serverIP, _serverPort, ^id(id responseObject, NSError* error, NSString* errMsg) {
            if (error)
            {
                [NSThread sleepForTimeInterval:0.6];
                retryTimes++;
                success = NO;
            }
            else
            {
                _socket = [responseObject intValue];
                success = YES;
            }
            
            return nil;
        }, -1);
        
        if (success)
        {
            _receiveDataThread = [[NSThread alloc] initWithTarget:self selector:@selector(receiveDataViaSocket:) object:nil];
            [_receiveDataThread start];
            
            if (!_timeoutCheckThread)
            {
                _timeoutCheckThread = [[NSThread alloc] initWithTarget:self selector:@selector(timeoutCheck:) object:nil];
                [_timeoutCheckThread start];
            }
            
            if (DataSocketStateDisconnecting != self.state)
            {
                self.state = DataSocketStateReady;
            }
            
            return;
        }
    }
    
    if (DataSocketStateDisconnecting != self.state)
    {
        self.state = DataSocketStateNotReady;
    }
}

- (void) close:(int)destState {
    @try
    {
        if (DataSocketStateNotReady != self.state)
        {
            if (_socket != 0)
            {
                disconnect(_socket, ^id(id responseObject, NSError *error, NSString *errMsg) {
                    _socket = 0;
                    return nil;
                });
            }
            
            [_cond lock];
            {
                while (_receiveDataThread)
                {
                    [_cond wait];
                }
                
                while (_timeoutCheckThread)
                {
                    [_cond wait];
                }
            }
            [_cond unlock];
        }
    }
    @catch (NSException *exception)
    {
        
    }
    @finally
    {
        self.state = (DataSocketState)destState;
        [self removeDataReceiver];
    }
}

- (void) timeoutCheck:(id)object {
    @try
    {
        int state = [self waitForState:(DataSocketStateNotReady | DataSocketStateReady)];
        if (DataSocketStateReady != state)
        {
            return;
        }
        
        while (YES)
        {
            [_cond lock];
            {
                while (!_currentDataReceiver && DataSocketStateReady == self.state)
                {
                    [_cond wait];
                }
                
                if (DataSocketStateReady != self.state)
                {
                    [_cond unlock];
                    return;
                }
                
                NSTimeInterval nowSeconds = [[NSDate date] timeIntervalSince1970];
                if (_currentDataReceiver.recentReceiveTime > 0
                    && nowSeconds * 1000 - _currentDataReceiver.recentReceiveTime > RECEIVE_TIMEOUT_LIMIT)
                {
                    NSLog(@"#DataConnectManager# TimeOut? _currentDataReceiver.recentReceiveTime=%lld, now=%lld", (long long)_currentDataReceiver.recentReceiveTime, (long long)(nowSeconds * 1000));
                    [self sendMessage:MSG_RECEIVE_TIMEOUT arg1:0 arg2:0 object:_currentDataReceiver];
                    
                    _currentDataReceiver = nil;
                    [self sendMessage:MSG_RECEIVER_EMPTIED arg1:0 arg2:0 object:nil];
                    
                    [_cond unlock];
                    return;
                }
            }
            [_cond unlock];
            
            [NSThread sleepForTimeInterval:5.0f];
        }
    }
    @catch (NSException *exception)
    {
        
    }
    @finally
    {
        [_cond lock];
        {
            _timeoutCheckThread = nil;
            [_cond broadcast];
        }
        [_cond unlock];
    }
}

- (void) receiveDataViaSocket:(id)object {
    int readSize = 0;
//    @try
//    {
        int state = [self waitForState:(DataSocketStateReady | DataSocketStateDisconnecting | DataSocketStateNotReady)];
        if (DataSocketStateReady != state)
        {
            //goto Finally;
            [_cond lock];
            {
                _receiveDataThread = nil;
                [_cond broadcast];
            }
            [_cond unlock];
            return;
        }
        
        do
        {
//            @try
//            {
                if ((readSize = (int)recv(_socket, _receiveBuffer, BUFFER_SIZE, 0)) <= 0)
                {
                    NSLog(@"Break on readSize = %d", readSize);
                    break;
                }
            //NSLog(@"Receive readSize = %d", readSize);
//            }
//            @catch (NSException *exception)
//            {
//                break;
//            }
//            @finally
//            {
//                break;
//            }
            
            [_cond lock];
            {
                if (_currentDataReceiver)
                {
                    [_currentDataReceiver didReceiveData:_receiveBuffer offset:0 length:readSize];
                    
                    if (_currentDataReceiver && _currentDataReceiver.isFinished)
                    {
                        [self removeDataReceiver];
                    }
                }
            }
            [_cond unlock];
        }
        while (readSize > 0);
        
        //            [_cond unlock];///!!!
        [self closeConnection];
        DataReceiver* receiver = [self removeDataReceiver];
        [self sendMessage:MSG_RECEIVE_ERROR arg1:0 arg2:0 object:receiver];
    
    goto Finally;
//    }
//    @catch (NSException *exception)
//    {
//        
//    }
//    @finally
Finally:
    {
        [_cond lock];
        {
            _receiveDataThread = nil;
            [_cond broadcast];
        }
        [_cond unlock];
    }
}

- (void) notifyStateChanged:(DataSocketState)newState oldState:(DataSocketState)oldState object:(id)object {
    [self sendMessage:MSG_STATE_CHANGED arg1:newState arg2:oldState object:object];
}

- (void) sendMessage:(int)what arg1:(int)arg1 arg2:(int)arg2 object:(id)object {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self handleMessage:what arg1:arg1 arg2:arg2 object:object];
    });
}

- (void) handleMessage:(int)what arg1:(int)arg1 arg2:(int)arg2 object:(id)object {
    switch (what)
    {
        case MSG_STATE_CHANGED:
        {
            [self onStateChanged:arg1 oldState:arg2 object:object];
        }
            break;
        case MSG_RECEIVE_ERROR:
        {
            DataReceiver* receiver = (DataReceiver*) object;
            if (receiver && [receiver respondsToSelector:@selector(onError:errMsg:)])
            {
                [receiver onError:DataSocketErrorException errMsg:@"SocketError"];
            }
        }
            break;
        case MSG_RECEIVE_TIMEOUT:
        {
            DataReceiver* receiver = (DataReceiver*) object;
            if (receiver && [receiver respondsToSelector:@selector(onError:errMsg:)])
            {
                [receiver onError:DataSocketErrorTimeout errMsg:@"ReceivingTimeOut"];
            }
        }
            break;
        case MSG_RECEIVER_EMPTIED:
        {
            [self onReceiverEmptied];
        }
            break;
        default:
            break;
    }
}

- (void) onStateChanged:(int)newState oldState:(int)oldState object:(id)object {
    for (id<DataConnectionObserver> observer in _observers)
    {
        if ([observer respondsToSelector:@selector(onDataConnectionStateChanged:oldState:object:)])
        {
            [observer onDataConnectionStateChanged:newState oldState:oldState object:object];
        }
    }
}

- (void) onReceiverEmptied {
    for (id<DataConnectionObserver> observer in _observers)
    {
        if ([observer respondsToSelector:@selector(onReceiverEmptied)])
        {
            [observer onReceiverEmptied];
        }
    }
}

- (int) socket {
    [self openConnection];
    if (DataSocketStateReady == [self waitForState:DataSocketStateReady])
        return _socket;
    else
        return 0;
}

@end

File: MADVCamera/CameraManager/DATAConnectManager.java
package com.madv360.madv.connection;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.ref.WeakReference;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketException;
import java.util.LinkedList;

import bootstrap.appContainer.ElephantApp;

/**
 * Created by wang yandong on 2016/5/24.
 */
public class DATAConnectManager {
    public static abstract class DataReceiver {
        protected abstract int onDataReceived(byte[] data, int offset, int length);

        protected abstract boolean isFinished();

        protected abstract void onError(int error, String errMsg);

        public void finish() {
            DATAConnectManager.getInstance().removeDataReceiver(this);
        }

        public DataReceiver() {
            recentReceiveTime = -1;
        }

        private long recentReceiveTime = -1;

        private int didReceiveData(byte[] data, int offset, int length) {
            recentReceiveTime = System.currentTimeMillis();
            return onDataReceived(data, offset, length);
        }
    }

    private static final String TAG = "QD:DataConnect";

    // 未连接
    public static final int SOCKET_STATE_NOT_READY = 1;
    // 开始连接server
    public static final int SOCKET_STATE_CONNECTING = 2;
    // 连接成功，正常发送接收中
    public static final int SOCKET_STATE_READY = 4;
    // 开始断开server
    public static final int SOCKET_STATE_DISCONNECTING = 8;

    public static final int ERROR_SOCKET_EXCEPTION = 1;
    public static final int ERROR_RECEIVING_TIMEOUT = 2;

    // Handler Messages :
    private static final int MSG_STATE_CHANGED = 10;
    private static final int MSG_RECEIVE_ERROR = 12;
    private static final int MSG_RECEIVE_TIMEOUT = 13;
    private static final int MSG_RECEIVER_EMPTIED = 14;

    //连接重试次数
    private static final int SOCKET_RETRY_TIMES_LIMIT = 5;
    private static final int RECEIVE_TIMEOUT_LIMIT = 15 * 1000;
    private static final int BUFFER_SIZE = 1024 * 1024;

    private static class DATAConnectManagerHolder {
        private static final DATAConnectManager INSTANCE = new DATAConnectManager();
    }

    public static final DATAConnectManager getInstance() {
        return DATAConnectManagerHolder.INSTANCE;
    }

    private DATAConnectManager() {
        Looper looper = Looper.getMainLooper();
        this.mCallbackHandler = new CallbackHandler(new WeakReference(this), looper);
        this.mContext = ElephantApp.getInstance().getApplicationContext();
        this.connectivityManager = (ConnectivityManager) mContext.getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    private DataReceiver currentDataReceiver = null;

    //socket连接回调
    private final Handler mCallbackHandler;

    private String serverIP = AMBACommands.AMBA_CAMERA_IP;
    private int serverPort = AMBACommands.AMBA_CAMERA_DATA_PORT;
    private int connTimeout = AMBACommands.AMBA_CAMERA_TIMEOUT;

    private int connState = SOCKET_STATE_NOT_READY;

    private Socket tcpSocket = null;
    private InputStream socketInputStream = null;
    private OutputStream socketOutputStream = null;
    private byte[] receiveBuffer = new byte[BUFFER_SIZE];

    private Thread timeoutCheckThread = null;
    private Thread receiveDataThread = null;
    private Context mContext;
    private ConnectivityManager connectivityManager;

    //回调对象
    private LinkedList<WeakReference<DataConnectionObserver>> observers = new LinkedList<>();

    public interface DataConnectionObserver {
        void onDataConnectionStateChanged(int newState, int oldState, Object object);

        void onReceiverEmptied();
    }

    public void addObserver(DataConnectionObserver observerReference) {
        this.observers.add(new WeakReference<>(observerReference));
    }

    public void removeObserver(DataConnectionObserver observerReference) {
        int index = 0;
        for (WeakReference<DataConnectionObserver> ref : observers) {
            if (ref.get() == observerReference) {
                break;
            }
            index++;
        }
        if (index < observers.size()) {
            observers.remove(index);
        }
    }

    public int getState() {
        return connState;
    }

    public void setState(int newState) {
        synchronized (this) {
            Log.d(TAG, "setState() from " + StringFromState(connState) + " to " + StringFromState(newState));
            int oldState = connState;
            connState = newState;
            notifyAll();

            notifyStateChanged(newState, oldState, this);
        }
    }

    public static final String StringFromState(int state) {
        switch (state) {
            case SOCKET_STATE_CONNECTING:
                return "SOCKET_STATE_CONNECTING";
            case SOCKET_STATE_DISCONNECTING:
                return "SOCKET_STATE_DISCONNECTING";
            case SOCKET_STATE_NOT_READY:
                return "SOCKET_STATE_NOT_READY";
            case SOCKET_STATE_READY:
                return "SOCKET_STATE_READY";
        }
        return "N/A";
    }

    public int waitForState(int state) {
        // Exit when: (state == 0 && connState == 0) || ((state & connState) !=
        synchronized (this) {
            while ((state != 0 || connState != 0) && ((state & connState) == 0)) {
                try {
                    wait();
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        }
        return connState;
    }

    public void notifyStateChanged(int newState, int oldState, Object object) {
        Message msg = mCallbackHandler.obtainMessage(MSG_STATE_CHANGED);
        msg.arg1 = newState;
        msg.arg2 = oldState;
        msg.obj = object;
        mCallbackHandler.sendMessage(msg);
    }

    public DataReceiver removeDataReceiver() {
        return removeDataReceiver(currentDataReceiver);
    }

    public DataReceiver removeDataReceiver(DataReceiver receiver) {
        DataReceiver ret = null;
        synchronized (this) {
            if (currentDataReceiver == receiver) {
                ret = currentDataReceiver;
                currentDataReceiver = null;
                notifyAll();

                Message msg = mCallbackHandler.obtainMessage(MSG_RECEIVER_EMPTIED);
                mCallbackHandler.sendMessage(msg);
            }
        }
        return ret;
    }

    public void setDataReceiver(DataReceiver receiver) {
        synchronized (this) {
            receiver.recentReceiveTime = System.currentTimeMillis();
            Log.e(TAG, "setDataReceiver(" + receiver + "): recentReceiveTime = " + receiver.recentReceiveTime);
            currentDataReceiver = receiver;
            notifyAll();
        }
    }

    public DataReceiver getDataReceiver() {
        return currentDataReceiver;
    }

    /**
     * 连接服务器
     */
    public void openConnection() {
        //建立socket连接
        int currentState = getState();
        if (currentState == SOCKET_STATE_CONNECTING || currentState == SOCKET_STATE_READY) {
            return;
        } else if (currentState == SOCKET_STATE_DISCONNECTING) {
            new Thread() {
                @Override
                public void run() {
                    waitForState(SOCKET_STATE_NOT_READY);
                    //重建连接
                    setState(SOCKET_STATE_CONNECTING);
                    open();
                }
            }.start();

        } else if (currentState == SOCKET_STATE_NOT_READY) {
            setState(SOCKET_STATE_CONNECTING);
            new Thread() {
                @Override
                public void run() {
                    open();
                }
            }.start();
        }
    }

    public void closeConnection() {
        int currentState = getState();
        if (currentState == SOCKET_STATE_DISCONNECTING || currentState == SOCKET_STATE_NOT_READY) {
            return;
        } else if (currentState == SOCKET_STATE_CONNECTING) {
            new Thread() {
                @Override
                public void run() {
                    waitForState(SOCKET_STATE_READY);
                    setState(SOCKET_STATE_DISCONNECTING);
                    close(SOCKET_STATE_NOT_READY);
                }
            }.start();
        } else if (currentState == SOCKET_STATE_READY) {
            setState(SOCKET_STATE_DISCONNECTING);
            new Thread() {
                @Override
                public void run() {
                    close(SOCKET_STATE_NOT_READY);
                }
            }.start();
        }
    }

    /**
     * 是否已经建立了连接
     */
    public boolean isConnected() {
        return ((getState() == SOCKET_STATE_READY)
                && (null != receiveDataThread && receiveDataThread.isAlive()));
    }

    /**
     * 重新连接服务器
     */
    private void open() {
        int retryTimes = 0;
        while (retryTimes < SOCKET_RETRY_TIMES_LIMIT
                && getState() != SOCKET_STATE_DISCONNECTING) {
            try {
                tcpSocket = new Socket();
                if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
                    Network wifiNetwork = null;
                    Network[] networks = this.connectivityManager.getAllNetworks();
                    for (Network network : networks) {
                        NetworkInfo networkInfo = this.connectivityManager.getNetworkInfo(network);
                        if (ConnectivityManager.TYPE_WIFI == networkInfo.getType()) {
                            wifiNetwork = network;
                            break;
                        }
                    }
                    if (null != wifiNetwork) {
                        wifiNetwork.bindSocket(tcpSocket);
                    }
                }
                tcpSocket.connect(new InetSocketAddress(serverIP, serverPort), connTimeout);
                socketInputStream = tcpSocket.getInputStream();
                socketOutputStream = tcpSocket.getOutputStream();

                receiveDataThread = new Thread(new ReceiveDataRunnable());
                receiveDataThread.start();

                if (null == timeoutCheckThread) {
                    timeoutCheckThread = new Thread(new TimeoutCheckRunnable());
                    timeoutCheckThread.start();
                }

                if (getState() != SOCKET_STATE_DISCONNECTING) {
                    //建立socket连接成功
                    setState(SOCKET_STATE_READY);
                }

                return;
            } catch (Exception e) {
                Log.e(TAG, "open : Exception : " + e.getMessage());
                try {
                    Thread.sleep(2000);
                } catch (InterruptedException e1) {
                    e1.printStackTrace();
                }
                retryTimes++;
                continue;
            }
        }

        if (getState() != SOCKET_STATE_DISCONNECTING) {
            //建立socket连接失败
            setState(SOCKET_STATE_NOT_READY);
        }
    }

    /**
     * 关闭当前连接
     */
    private void close(int destState) {
        try {
            if (getState() != SOCKET_STATE_NOT_READY) {
                //关闭socket对象
                try {
                    if (null != tcpSocket) {
                        tcpSocket.close();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                } finally {
                    tcpSocket = null;
                }

                //关闭输出流对象
                try {
                    if (null != socketOutputStream) {
                        socketOutputStream.close();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                } finally {
                    socketOutputStream = null;
                }

                //关闭输入流对象
                try {
                    if (null != socketInputStream) {
                        socketInputStream.close();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                } finally {
                    socketInputStream = null;
                }

//                //等待发送线程关闭
//                while (null != sendDataThread)
//                {
//                    try {
//                        wait();
//                    }
//                    catch (Exception ex) {
//                        ex.printStackTrace();
//                    }
//                }
                Log.v(TAG, "Before synchronized #0 @ DATAConnectManager$close");
                synchronized (this) {
                    Log.v(TAG, "Enter synchronized #0 @ DATAConnectManager$close");
                    //等待接收线程关闭
                    while (null != receiveDataThread) {
                        try {
                            wait();
                        } catch (Exception ex) {
                            ex.printStackTrace();
                        }
                    }

                    //等待超时检测线程关闭
                    while (null != timeoutCheckThread) {
                        try {
                            wait();
                        } catch (Exception ex) {
                            ex.printStackTrace();
                        }
                    }
                }
                Log.v(TAG, "After synchronized #0 @ DATAConnectManager$close");
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            setState(destState);
            removeDataReceiver();
        }
    }

    /**
     * 检测请求是否超时
     */
    private class TimeoutCheckRunnable implements Runnable {
        @Override
        public void run() {
            try {
                int state = waitForState(SOCKET_STATE_NOT_READY | SOCKET_STATE_READY);
                if (state != SOCKET_STATE_READY) {
                    return;
                }

                while (true) {
                    Log.v(TAG, "BEFORE synchronized @ TimeoutCheckRunnable#0");
                    synchronized (DATAConnectManager.this) {
                        Log.v(TAG, "ENTER synchronized @ TimeoutCheckRunnable#0");
                        while (currentDataReceiver == null && getState() == SOCKET_STATE_READY) {
                            try {
                                Log.e(TAG, "TimeoutCheckRunnable : Before wait");
                                DATAConnectManager.this.wait();
                            } catch (Exception ex) {
                                Log.e(TAG, "TimeoutCheckRunnable : Exception#0");
                                ex.printStackTrace();
                            }
                        }
                        Log.e(TAG, "TimeoutCheckRunnable : After wait");

                        if (getState() != SOCKET_STATE_READY) {
                            Log.e(TAG, "TimeoutCheckRunnable : Exit#0");
                            return;
                        }

                        if (currentDataReceiver.recentReceiveTime > 0 && System.currentTimeMillis() - currentDataReceiver.recentReceiveTime > RECEIVE_TIMEOUT_LIMIT) {
                            Log.e(TAG, "TimeoutCheckRunnable : Timeout on " + currentDataReceiver);
                            //closeConnection();

                            Message msg = mCallbackHandler.obtainMessage(MSG_RECEIVE_TIMEOUT, currentDataReceiver);
                            mCallbackHandler.sendMessage(msg);

                            currentDataReceiver = null;
                            msg = mCallbackHandler.obtainMessage(MSG_RECEIVER_EMPTIED);//Same effect as removeDataReceiver()
                            mCallbackHandler.sendMessage(msg);
                            Log.e(TAG, "TimeoutCheckRunnable : Exit#1");
                            return;
                        }
                    }
                    Log.v(TAG, "AFTER synchronized @ TimeoutCheckRunnable#0");
                    try {
                        Thread.sleep(5 * 1000);
                    } catch (InterruptedException ie) {
                        Log.e(TAG, "TimeoutCheckRunnable : Exception#1");
                        ie.printStackTrace();
                    }
                }
            } finally {
                Log.e(TAG, "TimeoutCheckRunnable : Finally");
                Log.v(TAG, "BEFORE synchronized @ TimeoutCheckRunnable#1");
                synchronized (DATAConnectManager.this) {
                    Log.v(TAG, "ENTER synchronized @ TimeoutCheckRunnable#1");
                    timeoutCheckThread = null;
                    DATAConnectManager.this.notifyAll();
                }
                Log.v(TAG, "AFTER synchronized @ TimeoutCheckRunnable#1");
            }
        }
    }

    public OutputStream getSocketOutputStream() {
        openConnection();
        if (DATAConnectManager.SOCKET_STATE_READY == waitForState(DATAConnectManager.SOCKET_STATE_READY))
            return socketOutputStream;
        else
            return null;
    }

    /**
     * 接收数据线程
     */
    private class ReceiveDataRunnable implements Runnable {
        public void run() {
            int readSize = 0;
            try {
                Log.d(TAG, "ReceiveDataRunnable : Before waitForState");
                int state = waitForState(SOCKET_STATE_READY | SOCKET_STATE_DISCONNECTING | SOCKET_STATE_NOT_READY);
                Log.d(TAG, "ReceiveDataRunnable : After waitForState");
                if (state != SOCKET_STATE_READY) {
                    Log.d(TAG, "ReceiveDataRunnable : Exit#0");
                    return;
                }

                do {
                    try {
                        if ((readSize = socketInputStream.read(receiveBuffer, 0, BUFFER_SIZE)) <= 0) {
                            break;
                        }
                    } catch (SocketException se) {
                        // 客户端主动socket.close()会调用这里
                        Log.d(TAG, "ReceiveDataRunnable : Exception#0 " + se.getMessage());
                        se.printStackTrace();
                        break;
                    } catch (IOException e) {
                        Log.d(TAG, "ReceiveDataRunnable : Exception#1 " + e.getMessage());
                        e.printStackTrace();
                        break;
                    } catch (Exception ex) {
                        Log.d(TAG, "ReceiveDataRunnable : Exception#2 " + ex.getMessage());
                        ex.printStackTrace();
                        break;
                    }

                    Log.v(TAG, "ReceiveDataRunnable : Socket read = " + readSize);
                    Log.v(TAG, "BEFORE synchronized @ ReceiveDataRunnable#0");
                    synchronized (DATAConnectManager.this) {
                        Log.v(TAG, "ENTER synchronized @ ReceiveDataRunnable#0");
                        if (null != currentDataReceiver) {
                            currentDataReceiver.didReceiveData(receiveBuffer, 0, readSize);

                            if (currentDataReceiver != null && currentDataReceiver.isFinished()) {
                                removeDataReceiver();///!!!Not Right!
                            }
                        }
                    }
                    Log.v(TAG, "AFTER synchronized @ ReceiveDataRunnable#0");
                }
                while (readSize > 0);

                Log.d(TAG, "ReceiveDataRunnable : finally#0 : readSize = " + readSize);
                closeConnection();
                DataReceiver receiver = removeDataReceiver();
                Message msg = mCallbackHandler.obtainMessage(MSG_RECEIVE_ERROR, receiver);
                mCallbackHandler.sendMessage(msg);
            } catch (Exception ex) {
                Log.d(TAG, "ReceiveDataRunnable : Exception#3 " + ex.getMessage());
                ex.printStackTrace();
            } finally {
                Log.d(TAG, "ReceiveDataRunnable : finally#1 ");
                Log.v(TAG, "BEFORE synchronized @ ReceiveDataRunnable#1");
                synchronized (DATAConnectManager.this) {
                    Log.d(TAG, "ENTER synchronized @ ReceiveDataRunnable#1");
                    receiveDataThread = null;
                    DATAConnectManager.this.notifyAll();
                }
                Log.v(TAG, "AFTER synchronized @ ReceiveDataRunnable#1");
            }
        }
    }

    private void onStateChanged(int newState, int oldState, Object object) {
        for (WeakReference<DataConnectionObserver> reference : observers) {
            DataConnectionObserver observer = reference.get();
            if (null != observer) {
                observer.onDataConnectionStateChanged(newState, oldState, object);
            }
        }
    }

    private void onReceiverEmptied() {
        for (WeakReference<DataConnectionObserver> reference : observers) {
            DataConnectionObserver observer = reference.get();
            if (null != observer) {
                observer.onReceiverEmptied();
            }
        }
    }

    private static class CallbackHandler extends Handler {
        private final WeakReference<DATAConnectManager> connectionManagerWeakReference;

        public CallbackHandler(WeakReference<DATAConnectManager> connectionManager, Looper looper) {
            super(looper);
            this.connectionManagerWeakReference = connectionManager;
        }

        public void handleMessage(Message msg) {
            DATAConnectManager connectionManager = this.connectionManagerWeakReference.get();
            if (null != connectionManager) {
                try {
                    switch (msg.what) {
                        case MSG_STATE_CHANGED:
                            connectionManager.onStateChanged(msg.arg1, msg.arg2, msg.obj);
                            break;
                        case MSG_RECEIVE_ERROR:
                            if (msg.obj != null) {
                                DataReceiver receiver = (DataReceiver) msg.obj;
                                receiver.onError(ERROR_SOCKET_EXCEPTION, "SocketError");
                            }
                            break;
                        case MSG_RECEIVE_TIMEOUT:
                            if (msg.obj != null) {
                                DataReceiver receiver = (DataReceiver) msg.obj;
                                receiver.onError(ERROR_RECEIVING_TIMEOUT, "ReceivingTimeOut");
                            }
                            break;
                        case MSG_RECEIVER_EMPTIED:
                            connectionManager.onReceiverEmptied();
                            break;
                        default:
                            break;
                    }
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        }
    }
}
